core/src/main/scala/filodb.core/memstore/ratelimit/CardinalityTracker.scala

```diff
@@ -46,6 +48,16 @@ class CardinalityTracker(ref: DatasetRef,
   // separates ws, ns, metric, etc. names
   val NAME_DELIMITER = ","
 
+  /**
+   * Map used to track cardinality count in downsample cluster.
+   * WHY this is used for Downsample cardinality count only ?
+   * This is because, in downsample cluster, we read through the entire index periodically and re-calculate
+   * the cardinality count from scratch. Without an in-memory aggregation data-structure, we would be modifying
+   * the records in RocksDB too frequently, causing a slowdown because of heavy disk writes.
+   * Hence we are using this map to help us aggregate in memory and then flush to RocksDB periodically
+   */
+  private val cardinalityCountMapDS : Map[Seq[String], (Int, Int)] = Map()
```

Comment:
No no to mutable member variables. We can not only use immutable structures with constructs like ``fold`` and ``reduce`` but pass those as parameters to flush.

Reply:
Sure @amolnayak311, could we have an offline sync on this quickly ? I wanted to understand this a bit for my understanding. 

While implementation, I was going through Scala collection doc and they mention the following and I wanted to avoid creating a new collection.

"Scala collections systematically distinguish between mutable and immutable collections. A mutable collection can be updated, reduced or extended in place. This means you can change, add, or remove elements of a collection as a side effect. Immutable collections, by contrast, never change. You have still operations that simulate additions, removals, or updates, but those operations will in each case return a new collection and leave the old collection unchanged."